### PR TITLE
lsp: Load config from parent dirs

### PR DESCRIPTION
**Pull request by charlieegan3**
Related https://github.com/StyraInc/regal/issues/626

https://github.com/StyraInc/regal/assets/1774239/d8108b47-568c-4ce9-a0fc-5fd608960b8e

### CODE REVIEW

**Comment by charlieegan3** [comment]
Huh, there seems to be something wrong with a flake. I'll need to look into that next week 😞

File: internal/lsp/config/watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		if err != nil {
+			t.Errorf("failed to start watcher: %v", err)
+		}
+	}()
+
+	watcher.Watch(configFilePath)
+
+	select {
+	case <-watcher.Reload:
+	case <-time.After(100 * time.Second):
```
Comment (anderseknert) [inline internal/lsp/config/watcher_test.go:42]: Can we cap this at like 5-10 seconds or something? Anything above that would seem like something we'd want to look into anyway.

File: internal/lsp/server.go [hunk 2]
```diff
@@ -434,7 +488,7 @@ func (l *LanguageServer) handleTextDocumentCodeAction(
 
 	for _, diag := range params.Context.Diagnostics {
 		switch diag.Code {
-		case "opa-fmt":
+		case ruleNameOPAFmt:
```
Comment (anderseknert) [inline internal/lsp/server.go:491]: We should do this for the v1 rule too, but I guess it's not rebase here?

**Review by anderseknert — APPROVED** [review]
A couple of questions but looks good! 

Do I understand correctly that this will scan for a config file in any directory above the workspace, but it's not possible to point out the config file explcitly, like from `arbitrary/dir/.regal/config.yaml`?

File: internal/lsp/config/watcher.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		w.path = opts.Path
+	}
+
+	return w
+}
+
+func (w *Watcher) Start(ctx context.Context) error {
+	var err error
+
+	err = w.Stop()
```
Comment (srenatus) [inline internal/lsp/config/watcher.go:46]: [nit] 
```suggestion
	err := w.Stop()
```

File: internal/lsp/config/watcher.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+	go func() {
+		w.loop(ctx)
+	}()
+
+	return nil
+}
+
+func (w *Watcher) loop(ctx context.Context) {
+	var err error
```
Comment (srenatus) [inline internal/lsp/config/watcher.go:64]: [style] Hmm there seems to be a pattern here.. there are a few situations where declaring this upfront is required, like
```go
foo := "y"
if x {
    foo, err = f() // := would introduce a new `foo` in this branch's scope, not what we want
}
```
but outside of those, I think I'd rather go with the `:=`..

**Review by srenatus — COMMENTED** [review]
Just a nitpick, sorry 🔍 😅

File: internal/lsp/config/watcher.go (hunk 4 above)
Comment (charlieegan3) [reply internal/lsp/config/watcher.go]: Thanks, that's fair. I have fixed in b4efd89

File: internal/lsp/config/watcher.go (hunk 3 above)
Comment (charlieegan3) [reply internal/lsp/config/watcher.go]: Also fixed in b4efd89

File: internal/lsp/server.go [hunk 5]
```diff
@@ -241,6 +246,55 @@ func (l *LanguageServer) formatToEdits(params ExecuteCommandParams, opts format.
...
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case path := <-l.configWatcher.Reload:
+			configFile, err := os.Open(path)
+			if err != nil {
+				l.logError(fmt.Errorf("failed to open config file: %w", err))
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/lsp/server.go:264]: I'll leave it for now 👍

File: internal/lsp/config/watcher_test.go (hunk 1 above)
Comment (charlieegan3) [reply internal/lsp/config/watcher_test.go]: Oops, that's a mistake. Fixed in d5bee69

File: internal/lsp/server.go (hunk 2 above)
Comment (charlieegan3) [reply internal/lsp/server.go:492]: I am lazy, golangci lint didn't flag that one... added in d5bee69

File: internal/lsp/config/watcher_test.go (hunk 1 above)
Comment (charlieegan3) [reply internal/lsp/config/watcher_test.go]: Made it 100ms like the others in bf48f11

Comment (anderseknert) [reply internal/lsp/config/watcher_test.go]: Thanks!

**Comment by charlieegan3** [comment]
I have made some adjustments to the server tests again to make them more robust to the order of messages. Feel free to take another look if you'd like or otherwise I think we can merge.

**Comment by anderseknert** [comment]
@charlieegan3 good to merge!

Just this question:

> Do I understand correctly that this will scan for a config file in any directory above the workspace, but it's not possible to point out the config file explcitly, like from `arbitrary/dir/.regal/config.yaml`?

But if this solves the problem as reported, I think that could be a future improvement if requested.

**Comment by charlieegan3** [comment]
I'll get back to the reporter and see if they can confirm 😊

**Pull request merged**